src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -2272,6 +2300,35 @@ class RGWDataSyncShardControlCR : public RGWBackoffControlCR {
   }
 };
 
+class RGWDataSyncShardNotifyCR : public RGWCoroutine {
+  RGWDataSyncEnv *sync_env;
```

Comment:
it looks like this is the same as `RGWMetaSyncShardNotifyCR` except for `RGWDataSyncEnv` vs `RGWMetaSyncEnv`. can you find a way to share this coroutine instead?